hamilton/plugins/jupyter_magic.py

```diff
@@ -14,26 +14,118 @@
 from IPython.display import HTML, display
 
-from hamilton import ad_hoc_utils, driver, lifecycle
+from hamilton import ad_hoc_utils, driver
+from hamilton.lifecycle import base as lifecycle_base
 
 
-def create_module(source: str, name: str = None) -> ModuleType:
-    """Create a temporary module from source code"""
-    module_name = ad_hoc_utils._generate_unique_temp_module_name() if name is None else name
-    module_object = ModuleType(module_name, "")
-    sys.modules[module_name] = module_object
-    exec(source, module_object.__dict__)
-    return module_object
+def get_assigned_variables(module_node: ast.Module) -> Set[str]:
+    """Get the set of variable names assigned in a AST Module"""
+    assigned_vars = set()
+
+    def visit_node(ast_node):
+        if isinstance(ast_node, ast.Assign):
+            for target in ast_node.targets:
+                if isinstance(target, ast.Name):
+                    assigned_vars.add(target.id)
+
+        for child_node in ast.iter_child_nodes(ast_node):
+            visit_node(child_node)
+
+    visit_node(module_node)
+    return assigned_vars
+
+
+def get_function_assigned_variables(code):
+    assigned_variables = []
+
+    def visit_Assign(node):
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                assigned_variables.append(target.id)
+
+    tree = ast.parse(code)
+    for node in ast.walk(tree):
+        if isinstance(node, ast.FunctionDef):
+            for sub_node in ast.walk(node):
+                if isinstance(sub_node, ast.Assign):
+                    visit_Assign(sub_node)
+
+    return assigned_variables
+
+
+def execute_and_get_assigned_values(shell: InteractiveShellApp, cell: str) -> Dict[str, Any]:
+    """Execute source code from a cell in the user namespace and collect
+    the values of all assigned variables into a dictionary.
+    """
+    shell.ex(cell)
+    expr = shell.input_transformer_manager.transform_cell(cell)
+    expr_ast = shell.compile.ast_parse(expr)
+    return {name: shell.user_ns[name] for name in get_assigned_variables(expr_ast)}
+
+
+def rebuild_driver(
+    dr: Optional[driver.Driver] = None,
+    config: Optional[Dict[str, Any]] = None,
+    modules: Optional[List[ModuleType]] = None,
+    adapters: Optional[
+        Union[lifecycle_base.LifecycleAdapter, List[lifecycle_base.LifecycleAdapter]]
+    ] = None,
+    graph_executor: Optional[driver.GraphExecutor] = None,
+    reload_modules: Literal[True, False, "strict"] = False,
+) -> driver.Driver:
+    _driver = dr if dr else driver.Builder().build()
+    _config = config if config else _driver.graph.config
+    _modules = modules if modules else _driver.graph_modules
+    _adapter = adapters if adapters else _driver.adapter
+    _graph_executor = graph_executor if graph_executor else _driver.graph_executor
+
+    if reload_modules:
+        new_modules = []
+        for m in _modules:
+            try:
+                new_module = importlib.reload(m)
```

Comment:
Does this work with the dynamic modules?

Reply:
My guess is it does but worth some comments as to how it interacts.

Reply:
Currently an unused feature, but added it since it came up to mind. The use case:

- User is writing dataflow in notebook
- User imports `module_x` and dynamics modules
- User updates `module_x` and wants it updated in the notebook without restarting the kernel.

User could manually `importlib.reload()` and rebuild the Driver, but we could automate it